### ISSUE TITLE
[Question] Chaining promises within sagas

### ISSUE THREAD

**Issue by hectorsabina** [redux-saga/redux-saga#1228]
So I've been struggling with this for a while now.

I've have a saga that does api requests. Looks like this (simplified):

```javascript
function* getRequest(url) {
  // prepare request
  // ...

  const myPromise = ((success, error) => {
    return fetch(url)
      .then(response => response.json())
      .catch(error => error)
  })

  return yield call(myPromise)
}
```

And the saga that invokes `getRequest`, looks like this:

```javascript
function* fetchData() {
  try {
    console.log('-- api request start')
    yield put(ActionCreators.apiRequestStart());
    const response = yield call(getRequest, 'path/to/endpoint');
    console.log('-- api success')
    yield put(ActionCreators.apiRequestSuccess());
    yield put(ActionCreators.updateData(response.data));
  } catch (e) {
    console.log('-- api error')
    yield put(ActionCreators.apiRequestError())
    yield put(ActionCreators.updateData(null));
  }
}
```

This *works* for me, but I don't understand why I need to do `yield call` twice. Also I wasn't able to get it working without wrapping the `fetch`call in another promise.

What I wanted to do is to have `getRequest` return the promise directly, and expected that the yielded call to this function (which returns a promise - since `fetch`returns a promise), will be waited for the middleware to deal with it and return the resolved value.

So basically, changing the `getRequest` method to this, doesn't work and I'd love to understand why :)

```javascript
 function getRequest(url) {
  // prepare request
  // ...

    return fetch(url)
      .then(response => response.json())
      .catch(error => error)
  })
}
```

**Comment by hectorsabina** [comment]
Also, forgot to mention that... if I do this in the `getRequest` method:

```javascript
 function* getRequest(url) {
  // prepare request
  // ...

    return yield call(fetch, url)
  })
}
```

That also works, but then I don't know how to chain the rest of methods (.then / .catch)

**Comment by Andarist** [comment]
This should work just fine
```js
function getRequest(url) {
  // prepare request
  // ...

    return fetch(url)
      .then(response => response.json())
      .catch(error => error)
  })
}
```

If it is not - you'd have to provide a repository with the bug reproduced. I'm pretty sure though this works as intended, using it my own project and it's covered with redux-saga's tests

**Comment by hectorsabina** [comment]
Hi @Andarist, thanks for your quick response. I actually figured out that the code you mention does work well for me too. I just realised the problem... I was defining `getRequest` as a generator function instead of a normal one, but I copied it differently in this issue somehow.

But it was helpful to know that it should work, now I'm on the right path.

Thanks!
